aptos-move/aptos-vm/src/keyless_validation.rs

```diff
@@ -185,30 +193,53 @@ pub(crate) fn validate_authenticators(
                                 )?;
+                            // println!("Public inputs hash time: {:?}", start.elapsed());
+
+                            let groth16_and_stmt =
+                                Groth16ProofAndStatement::new(*groth16proof, public_inputs_hash);
 
                             // The training wheels signature is only checked if a training wheels PK is set on chain
                             if training_wheels_pk.is_some() {
-                                zksig
-                                    .verify_training_wheels_sig(
-                                        training_wheels_pk.as_ref().unwrap(),
-                                        &public_inputs_hash,
-                                    )
-                                    .map_err(|_| {
-                                        invalid_signature!(
-                                            "Could not verify training wheels signature"
-                                        )
-                                    })?;
+                                match &zksig.training_wheels_signature {
+                                    Some(training_wheels_sig) => {
+                                        training_wheels_sig
+                                            .verify(
+                                                &groth16_and_stmt,
+                                                training_wheels_pk.as_ref().unwrap(),
+                                            )
+                                            .map_err(|_| {
+                                                invalid_signature!(
+                                                    "Could not verify training wheels signature"
+                                                )
+                                            })?;
+                                    },
+                                    None => {
+                                        return Err(invalid_signature!(
+                                            "Training wheels signature expected but it is missing"
+                                        ))
+                                    },
+                                }
                             }
 
-                            zksig
-                                .verify_groth16_proof(public_inputs_hash, pvk)
-                                .map_err(|_| invalid_signature!("Proof verification failed"))?;
+                            // If the ZKP has not successfully verified over this statment in the past,
+                            // manually verify it & cache it. Otherwise, skip.
+                            if matches!(ZKP_CACHE.get(&groth16_and_stmt), None | Some(false)) {
+                                let result = zksig.verify_groth16_proof(public_inputs_hash, pvk);
+
+                                if result.is_ok() {
+                                    ZKP_CACHE.insert(groth16_and_stmt, true);
+                                }
```

Comment:
why not also cache failed zkp_and_stmt?

Reply:
Hm. I guess we could. That cache would be very easy to pollute by sending bad proofs (which are cheap to create), so I wonder how useful it would actually be.